src/shims/unix/fs.rs

```diff
@@ -1385,15 +1389,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         Ok(Scalar::from_maybe_pointer(entry, this))
     }
 
-    fn macos_readdir_r(
+    fn macos_fbsd_readdir_r(
         &mut self,
         dirp_op: &OpTy<'tcx, Provenance>,
         entry_op: &OpTy<'tcx, Provenance>,
         result_op: &OpTy<'tcx, Provenance>,
     ) -> InterpResult<'tcx, Scalar<Provenance>> {
         let this = self.eval_context_mut();
 
-        this.assert_target_os("macos", "readdir_r");
+        if !matches!(&*this.tcx.sess.target.os, "macos" | "freebsd") {
+            panic!("macos_fbsd_readdir_r not meant for os {}", this.tcx.sess.target.os);
+        }
+
+        // FIXME?: oddily it seems to pick up the old libc crate's freebsd11 definition instead of freebsd13, should be 6
+        let name_offset = if this.tcx.sess.target.os == "macos" { 5 } else { 4 };
```

Comment:
Not sure why you mention libc here, this is a hard-coded offset, how should that pick up anything...?

Also, did that offset change between different versions of freebsd?!? How are we supposed to handle that, not knowing which version the program is built for?

Reply:
Oh wait, the name is just confusing. This is not an offset! It is a field index.

This should pick up the version of libc that comes with the standard library.

Reply:
And so the Rust libc crate then got that update as well at time point? Which libc crate version is that?

This [file here](https://github.com/rust-lang/rust/blob/91f7f266ce973a08e6e700f9ce815f7b61d347e9/Cargo.lock#L2167) should be defining which libc is picked up by Miri here. It's the latest version currently.

Reply:
in fact freebsd is one of the few oses in the libc's crate which maintains several os releases versions e.g. freebsd 11 up to 14 having their own definitions when there are changes.

Reply:
I don't quite follow, can you point me at the code for that in the libc crate? Which type does one get when asking for `libc::dirent`?

Reply:
Sure
[freebsd 11](https://github.com/rust-lang/libc/blob/main/src/unix/bsd/freebsdlike/freebsd/freebsd11/mod.rs#L215) and [later](https://github.com/rust-lang/libc/blob/main/src/unix/bsd/freebsdlike/freebsd/freebsd13/mod.rs#L234)

Reply:
Ah, I think I found it... [here](https://github.com/rust-lang/libc/blob/874e5277f968ca18420e121d1bd296cbefc47b1d/build.rs#L67). The libc build script runs some code to determine the FreeBSD version. When the FreeBSD standard library is built on a different OS (like e.g. when Miri CI tests the freebsd target on a Linux host), that will fail, and it defaults to freebsd11. But if you are using Miri on a freebsd host, it will use the host freebsd version.

Gosh, what a nightmare.^^ This is impossible to test properly.

Reply:
Okay so what this means for Miri is we have to support all versions of freebsd since we can't know what the build script picks. But we can only test one of them. Amazing. Maybe long-term we want to work with the libc crate to overwrite the version detection inside Miri, though maybe that has other issues.

For this PR, it means:
- definitely use `project_field_named` to get the name_place, that part should be easy
- for writing all the other fields, do a dynamic check whether the `d_off` field is present. We can't test on "is this freebsd or not" since the type is different for different versions of freebsd, we need to check the presence of this field. You will need to add a new helper function that checks whether a field of a given name exists.